### PR TITLE
ENH: lxt_ttc, txt, tt_delay / LaserTimingCompensation

### LINKED ISSUE #535: lxt_ttc
This motor moves both lxt and txt (I think?). See old python for more info

### PR DESCRIPTION
**Pull request by klauer**
Closes #535

Notes
-------
* `lxt_ttc`: time stage series compensation using stages `lxt` and `txt` (appears to be like a `SyncAxes`-style thing)
* `lxt` is already implemented as the `LaserTiming` class, a virtual motor based on vitara positions
* `txt` is a virtual motor based on `tt_delay` (`xppTTdelay`/`xcsTTdelay`), which uses delay stage to time (perhaps could use `DelayBase`) to convert the `tt_delay` stage position
* `tt_delay` is a motor record type `xps8p` (``pcdsdevices`` ``Newport``)

XCS
---

```python
lxt_ttc = pcdsdevices.lxe.LaserTimingCompensation(
    '',
    delay_prefix='XCS:LAS:MMN:01',
    laser_prefix='LAS:FS4',
    name='lxt_ttc',
)
```

XPP
---

```python
lxt_ttc = pcdsdevices.lxe.LaserTimingCompensation(
    '',
    delay_prefix='XPP:LAS:MMN:16',
    laser_prefix='LAS:FS3',
    name='lxt_ttc',
)
```

Additional notes:
* `lxt` - `LaserTiming` / Vitara
* `ttc` - stage that moves the time tool signal to keep
* `lxt_ttc` moves `lxt` and also `ttc` to compensate so that we can get the
  true laser x-ray delay by using the `lxt`-value and the result of the
  analysis of the time tool data and we don't count anything double. (Silke)
* `ttc` has a smaller range, so `lxt_ttc` will move both when it can. When
  `ttc` is out of range, it will only move `lxt` until we are back in the
  range where `ttc` can move. (Silke)

Conversion check
----------------

Note that the original ``xcsTTdelay`` conversion utilities don't take into
account the number of bounces and require external multiplication/division by
that.

TODO
-------
- [x] `lxt`
- [x] `tt_delay` (class `Newport`)
- [x] `txt` (`DelayNewport`, depends on `tt_delay`, relies on motor.EGU being set appropriately)
- [x] `lxt_ttc` (`SyncAxes`, depends on `lxt`, `txt`)
- [x] Capture additional notes from Silke + Zach in docstrings
- [x] `txt` unit conversion manual check
- [x] `txt` unit conversion check in testing
- [x] Invert `lxt` direction
- [x] Check if this does what it's supposed to
- [x] Add release notes
- [x] Tab whitelist
- [x] Fix tests after inversion change

### CODE REVIEW

**Comment by klauer** [comment]
I think this is largely there, with the notable exception of not being able to test it effectively.

Also, there is an important note:
The axes should be synchronized to move in the *opposite* directions, which this PR does not implement.
This is based on @ZryletTC 's indication that the direction of one of the motors may be swapped.

**Comment by klauer** [comment]
```
In [7]: lxt._fs_tgt_time.get()
Out[7]: -20999999999.999996

In [8]: lxt
Out[8]:

lxt
---
position: 20.999999999999996
setpoint: 20.999999999999996
```

File: pcdsdevices/lxe.py
Comment (ZryletTC) [inline pcdsdevices/lxe.py:219]: If it's to be specific for this application, maybe it could be named `_NegativeUnitConversionDerivedSignal` or `_LXTUnitConversionDerivedSignal`?
Or if it's meant to be more generic (maybe it could be expanded to accept scale as an init parameter in a future PR), maybe remove the `_` and move to signal.py?

File: pcdsdevices/lxe.py
Comment (ZryletTC) [inline pcdsdevices/lxe.py:230]: Isn't scale unitless?

**Review by ZryletTC — DISMISSED** [review]
Code looks good :)

File: pcdsdevices/lxe.py
Comment (klauer) [reply pcdsdevices/lxe.py:219]: 🤷‍♂️ I think keeping the leading underscore `_` leaves room for changing it in the future. I'll change it if you feel strongly enough.

File: pcdsdevices/lxe.py
Comment (klauer) [reply pcdsdevices/lxe.py:230]: Duh, yes. Thanks.

File: pcdsdevices/lxe.py
Comment (ZryletTC) [reply pcdsdevices/lxe.py:219]: True, keeping it private allows us to keep it as a WIP

**Comment by ZryletTC** [comment]
The SyncAxesBase changes are a little opaque to me at the moment as I haven't fully delved into how PseudoPositioners and RealPositions tick. I'll trust a passing test then try it out on hardware.

**Comment by klauer** [comment]
Still leaving this one unmerged for now - how to proceed hinges on whether or not the EPICS axis direction will be inverted:
1. This PR now includes the code to invert it in Python, but tests have not been fixed to account for it
2. There is a report that the axis direction may get inverted on the EPICS layer, obviating the need for the Python fix

Since (2) would make (1) unnecessary work, let's just wait a bit and see what happens...

**Comment by klauer** [comment]
OK!

**Comment by ZryletTC** [comment]
Looks good 👍 Will test on XCS laser setup

File: pcdsdevices/signal.py
Comment (ZLLentz) [inline pcdsdevices/signal.py:355]: Nitpicky, but can this really be `typing.Any`?

Comment (klauer) [reply pcdsdevices/signal.py:355]: `{float, int, np.int*, np.float*}` -> `numbers.Real`, I suppose is most accurate

**Comment by klauer** [comment]
There's a lot of inversion confusion at the beamline. @ZryletTC compiled the following:

```
Motor    | EPICS DIR | Python Reverse?
lxt      |   Neg     |       Yes
lxt_fast |   Neg     |       No
lxe      |   Neg     |       No
txt      |   Pos     |       No

lxt_ttc: lxt+ txt-
```

And this necessitated the reversal of `lxt` (_again_). Additional fix for lxt `set_current_position` in 1782f98.

**Comment by klauer** [comment]
<details>
<summary>No control limits on the vitara, only alarm limits:</summary>

```
(dev) [klauer@psbuild-rhel7-01  ~]$ caget -d 30 LAS:FS4:VIT:FS_TGT_TIME
LAS:FS4:VIT:FS_TGT_TIME
    Native data type: DBF_DOUBLE
    Request type:     DBR_CTRL_FLOAT
    Element count:    1
    Value:            4722.73
    Status:           NO_ALARM
    Severity:         NO_ALARM
    Units:            ns
    Precision:        6
    Lo disp limit:    0
    Hi disp limit:    0
    Lo alarm limit:   -200000
    Lo warn limit:    -20000
    Hi warn limit:    20000
    Hi alarm limit:   200000
    Lo ctrl limit:    0
    Hi ctrl limit:    0
```

PH motor has the following:
```
(dev) [klauer@psbuild-rhel7-01  ~]$ caget LAS:FS{4,5,6}:MMS:PH.LLM
LAS:FS4:MMS:PH.LLM             -1e+12
LAS:FS5:MMS:PH.LLM             -1e+12
LAS:FS6:MMS:PH.LLM             -1e+12
(dev) [klauer@psbuild-rhel7-01  ~]$ caget LAS:FS{4,5,6}:MMS:PH.HLM
LAS:FS4:MMS:PH.HLM             1e+12
LAS:FS5:MMS:PH.HLM             1e+12
LAS:FS6:MMS:PH.HLM             1e+12
```
</details>

Proof of reversal for `xcs_lxt_ttc` and not `lxt`:
```
In [4]: xcs_lxt_ttc.delay.position
Out[4]: _ReversedTimeToolDelayPseudoPos(delay=1.0000000000000003e-09)

In [5]: txt.position
Out[5]: TimeToolDelayPseudoPos(delay=-1.0000000000000003e-09)
```

**Comment by ZryletTC** [comment]
Before merging, can you please add `tab_whitelist` attributes to each of the user-facing lxe classes?

**Comment by klauer** [comment]
👍 Will do, @ZryletTC

**Comment by klauer** [comment]
Everything discussed above should be here now, I believe.

I think this'll be a target for rework down the line, considering how messy it ended up being.

**Comment by klauer** [comment]
OK, tests passing, new hack included and documented. Will merge later today if no further input.

**Comment by klauer** [comment]
🐰 🕳️

**Pull request merged**
